File: DailyQuestion/js/1162. AsFarfromLandasPossible.js
/** 
 * 输入：[[1,0,1],[0,0,0],[1,0,1]]
 * 输出：2
 * 解释： 
 * 海洋区域 (1, 1) 和所有陆地区域之间的距离都达到最大，最大距离为 2。
*/
let maxDistance = function(grid){
    let dx = [0,0,1,-1];
    let dy = [1,-1,0,0];
    let m = grid.length, n = grid[0].length;
    let arr = [];
    for(let i=0; i<m;i++){
        for(let j=0; j<n;j++){
            if(grid[i][j] === 1){
                arr.push([i,j]);
            }
        }
    }
    let hasOcean = false;
    let point = null;
    while(arr.length > 0){
        point = arr.shift();
        let x = point[0], y = point[1];
        for(let i=0;i<4;i++){
            let newX = x+dx[i];
            let newY = y+dy[i];
            if(newX < 0 || newX >= m || newY < 0 || newY >= n || grid[newX][newY] !== 0){
                continue;
            }
            grid[newX][newY] = grid[x][y]+1;
            hasOcean = true;
            arr.push([newX,newY]);
        }
    }
    if(point === null || !hasOcean){
        return -1;
    }
    return grid[point[0]][point[1]]-1;
};
console.log(maxDistance([[1,0,1],[0,0,0],[1,0,1]]));
console.log(maxDistance([[1,0,0],[0,0,0],[0,0,0]]));

File: DailyQuestion/java/1162. AsFarfromLandasPossible.java
package DailyQuestion.java;

import java.util.ArrayList;
import java.util.List;

class AsFarfromLandasPossible {
    public int maxDistance(int[][] grid) {
        int[] dx = {0, 0, 1, -1};
        int[] dy = {1, -1, 0, 0};
        List<int[]> list = new ArrayList<int[]>();
        int m = grid.length, n = grid[0].length;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j] == 1){
                    list.add(new int[]{i,j});
                }
            }
        }
        boolean hasOcean =false;
        int[] point=null;
        while(!list.isEmpty()){
            point = list.get(0);
            list.remove(0);
            int x=point[0], y=point[1];
            for(int i=0;i<4;i++){
                int newx=x+dx[i];
                int newy=y+dy[i];
                if(newx < 0 || newx >=m || newy < 0 || newy >= n || grid[newx][newy] !=0){
                    continue;
                }
                grid[newx][newy] = grid[x][y]+1;
                list.add(new int[]{newx,newy});
                hasOcean=true;
            }
        }
        if(point == null || !hasOcean){
            return -1;
        }
        return grid[point[0]][point[1]]-1;
    }
 
}